### DISCUSSION TITLE
Attempt to fix beatmaps with missing status

### DISCUSSION THREAD

**Discussion by Feodor0090** [ppy/osu#24311] [Q&A] [answered]
I'm not sure where the issue(s) actually is, so let me explain what i had, what i did and what i got.

I noticed some sets to lose their status at all (so, all difficulties are "none" and set itself is "none" too):

![изображение](https://github.com/ppy/osu/assets/53872073/e8b7dfa9-0b84-4f05-8450-2d8f842773f5)

Issue that sets may not detect upstream updates or not update their status is known and kinda wont-fixed, so i already used to ranked sets which are still "qualified" by lazer's thought and "pending" which are actually graveyarded, but losing status completely is something new, so i started to think what to do with it.

First attempt was, of course, to delete `online.db`. AFAIK it's filled by map updates info on each launch and the fact that part of updates are getting lost is also known.

It redownloaded and now it is 55 mb instead of 40 mb (15 mb of lines were lost somewhere, i guess?). This did not fixed anything - pending-while-actually-graveyarded maps actually turned into graveyarded, but qual-WA-ranked left qualed and more "none" maps have appeared. Specifically, sets which are wip/pending but all difficulties are "none":

![изображение](https://github.com/ppy/osu/assets/53872073/da19abb2-a260-4e21-a192-c5deae7ed926)

I remembered that last time when star rating was changed, all maps got reprocessed and many "frozen" quals changed their status to ranked. It was not too long to find `BackgroundBeatmapProcessor`, and first what i did was changing `if (ruleset.LastAppliedDifficultyVersion < currentVersion)` to `if (true)` in `checkForOutdatedStarRatings()`. Of course it did not work. It set all SR to -1 as expected, and then did nothing.
```
2023-07-20 15:43:50 [verbose]: Loading BackgroundBeatmapProcessor...
2023-07-20 15:43:50 [verbose]: Beginning background beatmap processing..
2023-07-20 15:43:50 [verbose]: Resetting star ratings for osu! (difficulty calculation version updated from 20220902 to 20220902)
2023-07-20 15:43:50 [verbose]: Loaded BackgroundBeatmapProcessor!
2023-07-20 15:43:50 [verbose]: 🌅 Global background loading
2023-07-20 15:43:50 [verbose]: Finished resetting 9393 beatmap sets for osu!
(other rulesets...)
2023-07-20 15:43:51 [verbose]: Resetting star ratings for hishigata (difficulty calculation version updated from 0 to 0)
2023-07-20 15:43:51 [verbose]: Finished resetting 0 beatmap sets for hishigata
2023-07-20 15:43:51 [verbose]: Querying for beatmap sets to reprocess...
2023-07-20 15:43:51 [verbose]: Finished background beatmap processing!
```

I added
```
if (t.Exception != null)
{
    Logger.Error(t.Exception, "Failed to reprocess");
}
```
to `ContinueWith` because only disposed exceptions were checked. I got NullRef from 129 line which was `beatmapSetIds.Add(b.BeatmapSet.ID);`, "guarded" by `Debug.Assert(b.BeatmapSet != null);`. Yeah, there were maps with null `BeatmapSet`- adding
```
if (b.BeatmapSet == null)
{
	Logger.Log($"{b} has null BMS");
	continue;
}
```
let it work. I got
```
2023-07-20 15:54:42 [verbose]: Akatsuki Records - A Melodys COLORS (Villdjack) [New Difficulty] has null BMS
2023-07-20 15:54:42 [verbose]: Akatsuki Records - A Melodys COLORS (Villdjack) [New Difficulty (1)] has null BMS
2023-07-20 15:54:43 [verbose]: YYY  - Waai WaiWai WaiWaiWai! (Feodor0090) [New Difficulty] has null BMS
```
in the logs but reprocessing worked. All "qualified" maps turned in ranked, finally, but now i have 50 beatmaps with lost status instead of 20 before reprocess:

![2023 07 20-19:15:06](https://github.com/ppy/osu/assets/53872073/3e631508-ef76-4757-92e1-b5bfb34cefe5)

1 and 2 "null" maps are created and deleted difficulties on graveyarded set. The 3rd is created-then-deleted difficulty on my own set too.

So, now i have list of issues/questions.

1) If player has broken beatmaps in DB, reprocess will silently fail without anything. I suggest adding at least
```diff
diff --git a/osu.Game/BackgroundBeatmapProcessor.cs b/osu.Game/BackgroundBeatmapProcessor.cs
index 9f62dae1f5..e422f47862 100644
--- a/osu.Game/BackgroundBeatmapProcessor.cs
+++ b/osu.Game/BackgroundBeatmapProcessor.cs
@@ -3,7 +3,6 @@
 
 using System;
 using System.Collections.Generic;
-using System.Diagnostics;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
@@ -64,9 +63,15 @@ protected override void LoadComplete()
                 convertLegacyTotalScoreToStandardised();
             }, TaskCreationOptions.LongRunning).ContinueWith(t =>
             {
-                if (t.Exception?.InnerException is ObjectDisposedException)
+                if (t.Exception != null)
                 {
-                    Logger.Log("Finished background aborted during shutdown");
+                    if (t.Exception.InnerException is ObjectDisposedException)
+                    {
+                        Logger.Log("Finished background aborted during shutdown");
+                        return;
+                    }
+
+                    Logger.Error(t.Exception, "Background beatmap processing failed!");
                     return;
                 }
 
@@ -120,7 +125,12 @@ private void processBeatmapSetsWithMissingMetrics()
             {
                 foreach (var b in r.All<BeatmapInfo>().Where(b => b.StarRating < 0 || (b.OnlineID > 0 && b.LastOnlineUpdate == null)))
                 {
-                    Debug.Assert(b.BeatmapSet != null);
+                    if (b.BeatmapSet == null)
+                    {
+                        Logger.Log($"{b} had null BMS during background reprocess!");
+                        continue;
+                    }
+
                     beatmapSetIds.Add(b.BeatmapSet.ID);
                 }
             });
```
so it would be visible.

2) Editor did not delete difficulties, but detached them from set and left as is somehow? There are only 3 of them so looks like flacky bug. I have no idea when (did that monthes ago) and how this happened.
3) Errors of metadata updater somehow may reset map's status at all. I suppose that it has different reason from status just not updating when it should, also, reprocess fixed all non-updated statuses while regressing "lost" statuses.
4) How do i fix this "none" beatmaps? I've run out of ideas. Some of them are just outdated, yeah, but others were not updated in years/monthes and i have latest version of them. Earlier i tried to delete this "others" and redownload them, but they lost their status after couple of game runs again. 
![изображение](https://github.com/ppy/osu/assets/53872073/4d16bc0e-0724-4ab7-ab86-9488b34c0be1)
This one, specifically, i tried to redownload 4 times but each game launch it turned from graveyard to none again, and it still is:
![изображение](https://github.com/ppy/osu/assets/53872073/587901c7-85f3-437f-bee2-b09b18ced412)

Logs from not-yet-finished reprocess just in case: 
[runtime (3-я копия).log](https://github.com/ppy/osu/files/12113532/runtime.3-.log)

**Comment by peppy** [comment, marked as answer]
I'm confused with what you're doing. star difficulty calculation has nothing to do with ranked status, which is fetched via `BeatmapUpdater` (`BeatmapUpdaterMetadataLookup`). There's already issues tracking edge cases where this doesn't have the correct status, which we'll eventually fix.
